### PR TITLE
[WIP] feat: `Builders.FromPath.LicenseEvidenceBuilder`

### LINKED ISSUE #1162: centralized license evidence gatrhering
Gathering license evidence is a common feature of downstream users.
As a library, this should be centralized in here, so that maintenance efforts are minimized.

currently known implementations downstream:
- for webpack: <https://github.com/CycloneDX/cyclonedx-webpack-plugin/blob/72700f06d00eac79fa3b91fe838bd78c583346a2/src/extractor.ts#L133-L173>
- for yarn: upcoming via <https://github.com/CycloneDX/cyclonedx-node-yarn/pull/193>

### PR DESCRIPTION
**Pull request by jkowalleck**
fixes #1162

> [!NOTE]
> on hold - see <https://github.com/CycloneDX/cyclonedx-javascript-library/pull/1158#issuecomment-2448604457>

----

refactored the evidence collection from our sister project <https://github.com/CycloneDX/cyclonedx-webpack-plugin>.

----

original: 
https://github.com/CycloneDX/cyclonedx-webpack-plugin/blob/72700f06d00eac79fa3b91fe838bd78c583346a2/src/extractor.ts#L133-L173

the original file is under Apache-2.0 license, and rights owner is OWASP.
All the same applies to this very project. Therefore, no legal issues exist -> we are free to copy/paste/modify the code without the slightest concerns.

----

need to see if there are additional requirements from https://github.com/CycloneDX/cyclonedx-node-yarn/pull/193

### CODE REVIEW

File: src/builders/fromPath.node.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    const contentType = getMimeForTextFile(file)
+    if (contentType === undefined) {
+      return undefined
+    }
+    let lname
+    if ( relativeFrom === undefined) {
+      lname = `file: ${file}`
+    } else {
+      // `file` could be absolute or relative path - lets resolve it anyway
+      file = resolve(relativeFrom, file)
```
Comment (jkowalleck) [inline src/builders/fromPath.node.ts:59]: unsure if this is a good idea ...

File: src/builders/fromPath.node.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   * Returns a generator for license evidences.
+   * Throws errors, if dir cannot be inspected.
+   *
+   * @param dir - path to inspect
+   * @param relativeFrom - path the dir shall be relative to
+   */
+  public * fromDir(dir: string, relativeFrom: string | undefined = undefined): Generator<NamedLicense> {
+    if ( relativeFrom !== undefined) {
+      // `dir` could be absolute or relative path - lets resolve it anyway
+      dir = resolve(relativeFrom, dir)
```
Comment (jkowalleck) [inline src/builders/fromPath.node.ts:86]: unsure if this is a good idea ...

**Comment by jkowalleck** [comment]
@AugustusKling,

this is the feature for license evidence builder.
Could I ask you for your opinion?

**Comment by AugustusKling** [comment]
@jkowalleck thanks for refactoring this to live in the library.

The code in this pull requests is can be consumed as long as packages are extracted into the file system before your code runs. This is not the case in the cyclonedx-node-yarn plugin which would mean it couldn't consume this part of the library.

Yarn supports 3 different linkers as explained in https://yarnpkg.com/features/linkers of which projects can freely choose:
* pnp: Uses a loader to make package content available at runtime. Does not extract packages to files on the file system which makes it impossible to access package contents with the Node file system API.
* pnpm: Uses a combination of symlinks and hard links as detailed in https://pnpm.io/symlinked-node-modules-structure. The Node file system API should work.
* node-modules: The copies as NPM would produce when writing its node_modules folder.

In other words, I think the `LicenseEvidenceBuilder` and also the `AttachmentFactory` should work for the pnpm and node-modules linkers. They would not work using the pnp linker which is essential in a plugin for Yarn.

In the cyclonedx-node-yarn plugin I need to use Yarn's file system abstraction (`FakeFS<PortablePath>`) which allows to see package contents regardless of the used linker and regardless if a project was ever installed. Doing so I could get the file names and the contents of the files to feed into the CDX library.

What do you think about exposing a `isLicenseFile(filename: string): boolean` function and one to create a license instance as in `makeLicense(filename: string, content: Buffer): CDX.Models.License`?

File: src/builders/fromPath.node.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        !this.#LICENSE_FILENAME_PATTERN.test(dci.name.toLowerCase())
+      ) {
+        continue
+      }
+
+      let le
+      try {
+        le = this.fromFile( join(dir, dci.name), relativeFrom)
+      } catch (e) {
+        continue
```
Comment (AugustusKling) [inline src/builders/fromPath.node.ts:103]: Since one would assume no error here, this code should fail in a clearly visible way. Either let the caller handle the error or log it **and** make sure the program terminates with a not-zero exit code.

File: src/factories/fromPath.node.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+   * Return an attachment on success.
+   * Returns undefined if it appears to be no known text file.
+   * Throws error, if content could not be fetched.
+   *
+   * Tries to guess the file's mime-type.
+   */
+  public fromTextFile(file: string): Attachment | undefined {
+    const contentType = getMimeTypeForTextFile(file)
+    if (contentType === undefined) {
+      return undefined
```
Comment (AugustusKling) [inline src/factories/fromPath.node.ts:63]: Finding something that matches the file pattern but no content type for it, should result in at least a warning.

**Comment by jkowalleck** [comment]
The work on this generalization with the goal to make it available in the library ... 
it feels like premature.... There is just not enough specific code in the field to actually produce any code worth shipping. 

I will postpone it until additional art exists.
